pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,484 @@
+}
+
+func (t *Transport) Start() error {
+	t.logger.Log(logging.LevelDebug, "mir handler is starting", "ownID", t.ownID, "listen", t.host.Addrs())
+	t.host.SetStreamHandler(ProtocolID, t.mirHandler)
+	return nil
+}
+
+func (t *Transport) Stop() {
+	t.logger.Log(logging.LevelDebug, "Stopping libp2p transport.", "ownID", t.ownID)
+	defer t.logger.Log(logging.LevelDebug, "Stopping libp2p transport finished.", "ownID", t.ownID)
+
+	t.host.RemoveStreamHandler(ProtocolID)
+
+	if err := t.host.Close(); err != nil {
+		t.logger.Log(logging.LevelError, "Could not close libp2p", "ownID", t.ownID, "err", err)
+	} else {
+		t.logger.Log(logging.LevelDebug, "libp2p host closed", "ownID", t.ownID)
+	}
+
+	t.connWg.Wait()
+}
+
+func (t *Transport) CloseOldConnections(newNodes map[types.NodeID]types.NodeAddress) {
+	t.connWg.Add(1)
+
+	go func() {
+		defer t.connWg.Done()
+
+		t.nodesLock.Lock()
+		defer t.nodesLock.Unlock()
+
+		for nodeID, dest := range t.nodes {
+			// Close an old connection to a node if we don't need to connect to it further.
+			if _, foundInNewNodes := newNodes[nodeID]; !foundInNewNodes {
+				t.logger.Log(logging.LevelDebug, "Closing old connection", "src", t.ownID, "dst", nodeID)
+
+				if err := t.host.Network().ClosePeer(dest.AddrInfo.ID); err != nil {
+					t.logger.Log(logging.LevelError, "Could not close old connection to node", "src", t.ownID, "dst", nodeID, "err", err)
+					continue
+				}
+
+				delete(t.nodes, nodeID)
+			}
+		}
+	}()
+}
+
+func (t *Transport) Connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	if len(nodes) == 0 {
+		t.logger.Log(logging.LevelWarn, "no nodes to connect to")
+		return
+	}
+
+	t.nodesLock.Lock()
+	for nodeID, addr := range nodes {
+		if nodeID == t.ownID {
+			continue
+		}
+		_, found := t.nodes[nodeID]
+		if !found {
+			info, err := peer.AddrInfoFromP2pAddr(addr)
+			if err != nil {
+				t.logger.Log(logging.LevelWarn, "connect: failed to parse addr", "src", t.ownID, "dest", nodeID, "addr", addr, "err", err)
+				continue
+			}
+			t.nodes[nodeID] = &nodeInfo{
+				ID:        nodeID,
+				Addr:      addr,
+				AddrInfo:  info,
+				IsOpening: false,
+				Stream:    nil,
+			}
+
+		}
+	}
+	t.nodesLock.Unlock()
+
+	t.connect(ctx, nodes)
+}
+
+func (t *Transport) connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	t.connWg.Add(1)
+	defer t.connWg.Done()
+
+	go func() {
+		t.logger.Log(logging.LevelDebug, "started connecting nodes", "src", t.ownID)
+		defer t.logger.Log(logging.LevelDebug, "finished connecting nodes", "src", t.ownID)
+
+		wg := &sync.WaitGroup{}
+		wg.Add(len(nodes))
+
+		for nodeID := range nodes {
+			if nodeID == t.ownID {
+				// Do not establish a real connection with own node.
+				wg.Done()
+				continue
+			}
+
+			if t.streamExists(nodeID) {
+				t.logger.Log(logging.LevelInfo, "stream to node already exists", "src", t.ownID, "dst", nodeID)
+				wg.Done()
+				continue
+			}
+
+			go t.connectToNode(ctx, nodeID, wg)
```

Comment:
`connectToNode` doesn't seem to re-check if the stream is already exist with `nodesLock` held. I think there is a possible race condition.

Reply:
Do you want to check that again?

Reply:
Not necessarily, we just need to make sure no race condition happens here.